[Filters] Calculating the CSSFilter geometry and clipping sometimes is incorrect

shallawa opened the pull request:
#### 7acdd2650a9bc80317594614894d344b16ae6a16
<pre>
[Filters] Calculating the CSSFilter geometry and clipping sometimes is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=254062">https://bugs.webkit.org/show_bug.cgi?id=254062</a>
rdar://106844401

Reviewed by Simon Fraser.

Like SVGFilter, CSSFilter should always intersect the FilterEffect imageRect with
its primitiveSubregion. This will require:

1. Filter::ClipOperation will be removed.
2. Filter::clipToMaxEffectRect() will always clip the rectangle with maxEffectRect.
3. FilterEffect::calculateImageRect() will return the maxEffectRect if number of
   input effects is zero. This was a bug which was covered by the union mode of
   Filter::clipToMaxEffectRect().
4. RenderLayerFilters::beginFilterEffect() should set its targetBoundingBox the
   sourceImageRect of FilterTargetSwitcher.
5. SVGFilterGraph::getNamedNodes() should not return an error if one of the inputs
   is builtin and there is no node for it. This happens when we build a graph of
   primitives and one of the inputs is SourceGraphic or SourceAlpha. In this case
   we just need to ignore this input since there is no built-in primitive for it.
   This fixes the calculation of the filter outsets which was covered by the union
   mode of Filter::clipToMaxEffectRect().
6. SVGFEMorphologyElement::outsets() needs to implemented. This is a bug which was
   covered by the union mode of Filter::clipToMaxEffectRect().

Note: This work is towards enabling the GraphicsContext filters. The clipping has
to fixed first since no intermidate buffers are used to apply the CSSFilter.

* LayoutTests/css3/filters/effect-blur-hw.html:
* LayoutTests/css3/filters/effect-blur.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/svg-morphology-clipped-expected.html: Added.
* LayoutTests/css3/filters/svg-morphology-clipped.html: Added.
* LayoutTests/fast/filter-image/filter-image-blur-expected.html:
* LayoutTests/fast/filter-image/filter-image-blur.html:
* LayoutTests/fast/filter-image/resources/svg-blur.svg:
* LayoutTests/fast/filter-image/resources/svg-noblur.svg:
* LayoutTests/fast/hidpi/filters-drop-shadow-expected.html: Added.
* LayoutTests/fast/hidpi/filters-drop-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/filter-region-transformed-child-001-expected.html:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
(WebCore::Filter::clipToMaxEffectRect const):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::Filter):
(WebCore::Filter::clipOperation const): Deleted.
(WebCore::Filter::setClipOperation): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::calculateImageRect const):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createBlurEffect):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::buildFilterFunctions):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(WebCore::SVGFEMorphologyElement::outsets const):
* Source/WebCore/svg/SVGFEMorphologyElement.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::SVGFilterGraph):
(WebCore::SVGFilterGraph::sourceGraphic const):
(WebCore::SVGFilterGraph::sourceAlpha const):
(WebCore::SVGFilterGraph::addNamedNode):
(WebCore::SVGFilterGraph::getNamedNode const):
(WebCore::SVGFilterGraph::getNamedNodes const):
(WebCore::SVGFilterGraph::isSourceName):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;Filter&gt;::encode):
(IPC::ArgumentCoder&lt;Filter&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/261827@main">https://commits.webkit.org/261827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fded22759800018adff67dacca8dd224bdae458

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121471 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13305 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5967 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118773 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17450 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100710 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106076 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99408 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1246 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46479 "1 flakes 139 failures 1 missing results") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14434 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1288 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15143 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10622 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20449 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53279 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8258 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16992 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
```diff
@@ -129,6 +128,11 @@ class SVGFilterGraph {
     }
 
 private:
+    static bool isBuiltName(const AtomString& id)
```
Did you mean "built-in"? Maybe "isSource"?

shallawa committed:
[Filters] Calculating the CSSFilter geometry and clipping sometimes is incorrect
https://bugs.webkit.org/show_bug.cgi?id=254062
rdar://106844401

Reviewed by Simon Fraser.

Like SVGFilter, CSSFilter should always intersect the FilterEffect imageRect with
its primitiveSubregion. This will require:

1. Filter::ClipOperation will be removed.
2. Filter::clipToMaxEffectRect() will always clip the rectangle with maxEffectRect.
3. FilterEffect::calculateImageRect() will return the maxEffectRect if number of
   input effects is zero. This was a bug which was covered by the union mode of
   Filter::clipToMaxEffectRect().
4. RenderLayerFilters::beginFilterEffect() should set its targetBoundingBox the
   sourceImageRect of FilterTargetSwitcher.
5. SVGFilterGraph::getNamedNodes() should not return an error if one of the inputs
   is builtin and there is no node for it. This happens when we build a graph of
   primitives and one of the inputs is SourceGraphic or SourceAlpha. In this case
   we just need to ignore this input since there is no built-in primitive for it.
   This fixes the calculation of the filter outsets which was covered by the union
   mode of Filter::clipToMaxEffectRect().
6. SVGFEMorphologyElement::outsets() needs to implemented. This is a bug which was
   covered by the union mode of Filter::clipToMaxEffectRect().

Note: This work is towards enabling the GraphicsContext filters. The clipping has
to fixed first since no intermidate buffers are used to apply the CSSFilter.

* LayoutTests/css3/filters/effect-blur-hw.html:
* LayoutTests/css3/filters/effect-blur.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/svg-morphology-clipped-expected.html: Added.
* LayoutTests/css3/filters/svg-morphology-clipped.html: Added.
* LayoutTests/fast/filter-image/filter-image-blur-expected.html:
* LayoutTests/fast/filter-image/filter-image-blur.html:
* LayoutTests/fast/filter-image/resources/svg-blur.svg:
* LayoutTests/fast/filter-image/resources/svg-noblur.svg:
* LayoutTests/fast/hidpi/filters-drop-shadow-expected.html: Added.
* LayoutTests/fast/hidpi/filters-drop-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/filter-region-transformed-child-001-expected.html:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
(WebCore::Filter::clipToMaxEffectRect const):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::Filter):
(WebCore::Filter::clipOperation const): Deleted.
(WebCore::Filter::setClipOperation): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::calculateImageRect const):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createBlurEffect):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::buildFilterFunctions):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(WebCore::SVGFEMorphologyElement::outsets const):
* Source/WebCore/svg/SVGFEMorphologyElement.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::SVGFilterGraph):
(WebCore::SVGFilterGraph::sourceGraphic const):
(WebCore::SVGFilterGraph::sourceAlpha const):
(WebCore::SVGFilterGraph::addNamedNode):
(WebCore::SVGFilterGraph::getNamedNode const):
(WebCore::SVGFilterGraph::getNamedNodes const):
(WebCore::SVGFilterGraph::isSourceName):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<Filter>::encode):
(IPC::ArgumentCoder<Filter>::decode):

Canonical link: https://commits.webkit.org/261827@main
